apps/mobile/src/screens/Login/EnterEmailScreen.tsx

```diff
@@ -24,48 +27,81 @@ export function EnterEmailScreen() {
   const [isLoggingIn, setIsLoggingIn] = useState(false);
 
   const [login] = useLogin();
+  const [verifyEmail] = useVerifyEmailMagicLink();
   const reportError = useReportError();
   const track = useTrack();
 
   const handleContinue = useCallback(async () => {
+    let hasNavigatedForward = false;
+
     setError('');
     setIsLoggingIn(true);
 
-    // Show the waiting screen.
-    navigation.navigate('WaitingForConfirmation', { email });
-
-    function handleLoginError(message: string) {
-      reportError(`LoginError: ${message}`);
+    function handleLoginError({
+      message,
+      underlyingError,
+    }: {
+      message: string;
+      underlyingError?: Error;
+    }) {
+      if (underlyingError) {
+        reportError(underlyingError);
+      } else {
+        reportError(`LoginError: ${message}`);
+      }
 
       setError(message);
 
-      // Bring the user back to the login route with an error displayed.
-      navigation.goBack();
+      if (hasNavigatedForward) {
+        navigation.goBack();
+      }
     }
 
     try {
+      const { verifyEmailMagicLink } = await verifyEmail({ variables: { input: { email } } });
+
+      if (verifyEmailMagicLink?.__typename === 'ErrInvalidInput') {
+        return handleLoginError({
+          message: `The email address you entered doesn't look valid. Please try again.`,
+        });
+      } else if (verifyEmailMagicLink?.__typename === 'VerifyEmailMagicLinkPayload') {
+        if (!verifyEmailMagicLink.canSend) {
+          return handleLoginError({
+            message: `We don't recognize the email you entered. Login with QR code instead, or verify your email at gallery.so/settings`,
+          });
+        }
+      } else {
+        return handleLoginError({ message: FALLBACK_ERROR_MESSAGE });
+      }
+
+      // Show the waiting screen.
+      hasNavigatedForward = true;
+      navigation.navigate('WaitingForConfirmation', { email });
+
       const token = await magic.auth.loginWithMagicLink({ email, showUI: false });
 
       if (!token) {
-        handleLoginError(
-          "Something unexpected went wrong while logigng in. We've been notified and are looking into it"
-        );
-        return;
+        return handleLoginError({ message: FALLBACK_ERROR_MESSAGE });
       }
 
       const result = await login({ magicLink: { token } });
 
       if (result.kind === 'failure') {
         track('Sign In Failure', { 'Sign in method': 'Email', error: result.message });
```

Comment:
let's make sure this is called in earlier cases of failure / early returns. maybe just roll into `handleLoginError`.

Reply:
great callout